### PR TITLE
feat: prevent csrf attacks with OAuth2.0 state

### PR DESCRIPTION
**Pull request by RyaoChengfeng**
## What's changed

Fix: [#867](https://github.com/casdoor/casdoor/issues/867) 
Fix: [#504](https://github.com/casdoor/casdoor/issues/504) 
Fix: [#564](https://github.com/casdoor/casdoor/issues/564)

1. The frontend gets a random `state` value by requesting `/api/get-oauth-state` and store it in cookies. When send a request to the idp, the state parameter will be added to the request. 
2. After the request is sent, the user is redirected back to the application. The `state` value will be included in this redirect. 
3. Retrieve the returned `state` value and compare it with the one you stored earlier. If the values match, then approve the authentication response and delete `state` in cookies.

### CODE REVIEW

**Comment by hsluoyz** [comment]
@Abingcbc @seriouszyx @Resulte plz review

**Comment by fanux** [comment]
[sealos](https://github.com/labring/sealos) needs this feature~ : )

File: web/src/auth/LoginPage.js [hunk 1]
```diff
@@ -78,6 +79,7 @@ class LoginPage extends React.Component {
   }
 
   UNSAFE_componentWillMount() {
+    this.getOAuthState();
```
[the remark this answers is not in the document]
Comment (RyaoChengfeng) [reply web/src/auth/LoginPage.js]: It is only used in oauth, but `href` in the <a></a> tag will call `Provider.getAuthUrl` function when it is rendered, and the state value needs to be obtained before this.

You can see `LoginPage.js` Line 349 and Line 364.

**Comment by hsluoyz** [comment]
@Abingcbc @seriouszyx plz review again

@fanux can you try this PR to see if it works?

**Review by seriouszyx — APPROVED** [review]
LGTM

**Comment by Abingcbc** [comment]
LGTM

File: controllers/auth.go [hunk 2]
```diff
@@ -314,10 +313,11 @@ func (c *ApiController) Login() {
 
 			setHttpClient(idProvider, provider.Type)
 
-			if form.State != conf.GetConfigString("authState") && form.State != application.Name {
-				c.ResponseError(fmt.Sprintf("state expected: \"%s\", but got: \"%s\"", conf.GetConfigString("authState"), form.State))
```
Comment (hsluoyz) [inline controllers/auth.go:318]: Is `authState` in app.conf not used anywhere after this change? Is yes, plz delete it everywhere

image

File: go.sum [hunk 3]
```diff
@@ -232,8 +232,9 @@ github.com/google/pprof v0.0.0-20200708004538-1a94d8640e99/go.mod h1:ZgVRPoUq/hf
 github.com/google/pprof v0.0.0-20200905233945-acf8798be1f7/go.mod h1:ZgVRPoUq/hfqzAqh7sHMqb3I9Rq5C59dIz2SbBwJ4eM=
 github.com/google/renameio v0.1.0/go.mod h1:KWCgfxg9yswjAJkECMjeO8J8rahYeXnNhOm40UhjYkI=
 github.com/google/uuid v1.1.2/go.mod h1:TIyPZe4MgqvfeYDBFedMoGGpEw/LqOeaOT+nhxU+yHo=
```
Comment (hsluoyz) [inline go.sum:234]: Did you run `go mod tidy`?

File: go.mod [hunk 4]
```diff
@@ -20,7 +20,7 @@ require (
 	github.com/go-pay/gopay v1.5.72
 	github.com/go-sql-driver/mysql v1.5.0
 	github.com/golang-jwt/jwt/v4 v4.2.0
-	github.com/google/uuid v1.2.0
+	github.com/google/uuid v1.3.0
```
Comment (hsluoyz) [inline go.mod:23]: Why update the version?

File: controllers/auth.go [hunk 5]
```diff
@@ -506,3 +506,13 @@ func (c *ApiController) HandleSamlLogin() {
 		slice[4], relayState, samlResponse)
 	c.Redirect(targetUrl, 303)
 }
+
+func (c *ApiController) GetOAuthState() {
+	state := strings.ReplaceAll(util.GenerateUuid(), "-", "")
+	c.Ctx.SetCookie("oauth_state", state)
+	c.Data["json"] = map[string]string{
+		"state": state,
```
Comment (hsluoyz) [inline controllers/auth.go:514]: Why not use `ResponseOK()`?

File: util/random.go [hunk 6]
```diff
@@ -23,3 +26,7 @@ func GenerateClientId() string {
 func GenerateClientSecret() string {
 	return randstr.Hex(20)
 }
+
+func GenerateUuid() string {
```
Comment (hsluoyz) [inline util/random.go:30]: Why don't re-use `GenerateClientId` or `GenerateClientSecret`?

File: web/src/auth/LoginPage.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (hsluoyz) [reply web/src/auth/LoginPage.js:82]: @Abingcbc

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: go.sum (hunk 3 above)
Comment (RyaoChengfeng) [reply go.sum:234]: Yes

Comment (hsluoyz) [reply go.sum:234]: OK

**Comment by hsluoyz** [comment]
Already completed

**Pull request closed without merging**